bug: Stencil should not require puppeteer dependency in new projects

### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/main/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/main/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already report this problem, without success.

### Stencil Version

v4

### Current Behavior

When creating a new project and removing the `puppeteer` dev dependency, Stencil will fail to start with the following exception:

```ts
[ ERROR ]  TypeScript: node_modules/@stencil/core/testing/jest/jest-environment.d.ts:9:44
           Cannot find module 'puppeteer' or its corresponding type declarations.

      L8:  setup(): Promise<void>;
      L9:  newPuppeteerPage(): Promise<import("puppeteer").Page>;
     L10:  closeOpenPages(): Promise<void>;

[ ERROR ]  TypeScript: node_modules/@stencil/core/testing/puppeteer/puppeteer-declarations.d.ts:3:120
           Cannot find module 'puppeteer' or its corresponding type declarations.

      L2:  import type { EventInitDict, EventSpy, ScreenshotDiff, ScreenshotOptions } from '@stencil/core/internal';
      L3:  ions, HTTPResponse, Page, ScreenshotOptions as PuppeteerScreenshotOptions, WaitForOptions } from 'puppeteer';
      L4:  /**

[27:32.9]  build failed, watching for changes... in 4.68 s
```

### Expected Behavior

Stencil should not require puppeteer as a dependency. Stencil projects may choose to use Playwright to test their component libraries and currently will require the dev dependency to satisfy Stencil. 

This worked previously in v2. Unsure if it regressed in v3 or v4. 

### System Info

```shell
System: node 16.15.1
    Platform: darwin (22.2.0)
   CPU Model: Apple M1 (8 cpus)
    Compiler: /Users/sean/Documents/ionic/issues/stencil-without-puppeteer/node_modules/@stencil/core/compiler/stencil.js
       Build: 1687953951
     Stencil: 4.0.1 ⛹
  TypeScript: 5.0.4
      Rollup: 2.42.3
      Parse5: 7.1.2
      Sizzle: 2.42.3
      Terser: 5.18.1
```


### Steps to Reproduce

1. Create a new project: `npm init stencil`
2. Remove the `puppeteer` dev dependency
3. Install dependencies
4. Run `npm run start`
5. Observe: Exception is thrown

### Code Reproduction URL

https://github.com/sean-perkins/stencil-puppeteer-bug

### Additional Information

_No response_

--- a/scripts/bundles/testing.ts
+++ b/scripts/bundles/testing.ts
@@ -106,6 +106,7 @@ export async function testing(opts: BuildOptions) {
         preferConst: true,
       }),
       prettyMinifyPlugin(opts, getBanner(opts, `Stencil Testing`, true)),
+      ignorePuppteerDependency(opts),
     ],
     treeshake: {
       moduleSideEffects: false,
@@ -129,4 +130,30 @@ async function copyTestingInternalDts(opts: BuildOptions, inputDir: string) {
       return false;
     },
   });
+}
+
+/**
+ * To avoid having user to install puppeteer for building their app (even if they don't use e2e testing),
+ * we ignore the puppeteer dependency in the generated d.ts file.
+ * @param opts build options
+ * @returns void
+ */
+function ignorePuppteerDependency(opts: BuildOptions) {
+  return {
+    name: 'ignorePuppteerDependency',
+    async buildEnd() {
+      const typeFilePath = join(opts.output.testingDir, 'puppeteer', 'puppeteer-declarations.d.ts');
+      const updatedFileContent = (await fs.readFile(typeFilePath, 'utf8'))
+        .split('\n')
+        .reduce((lines, line) => {
+          if (line.endsWith(`from 'puppeteer';`)) {
+            lines.push('// @ts-ignore - avoid requiring puppeteer as dependency');
+          }
+          lines.push(line);
+          return lines;
+        }, [] as string[])
+        .join('\n');
+      await fs.writeFile(typeFilePath, updatedFileContent);
+    },
+  };
 }

